File: 09-intervals/car-pooling/solution.cpp
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;

class Solution {
public:
    bool carPooling(const vector<vector<int>>& trips, int capacity) {
        vector<pair<int, int>> events;
        events.reserve(trips.size() * 2);

        for (const auto& trip : trips) {
            const int passengers = trip[0];
            const int start = trip[1];
            const int end = trip[2];

            events.push_back({start, passengers});
            events.push_back({end, -passengers});
        }

        // Sort by location, then process drop-offs before pick-ups at the same stop.
        sort(events.begin(), events.end());

        int currentLoad = 0;
        for (const auto& [location, change] : events) {
            (void)location;
            currentLoad += change;
            if (currentLoad > capacity) {
                return false;
            }
        }

        return true;
    }
};

int main() {
    Solution solution;

    vector<vector<int>> trips = {{2, 1, 5}, {3, 3, 7}};

    cout << boolalpha;
    cout << "capacity = 4 -> " << solution.carPooling(trips, 4) << '\n';
    cout << "capacity = 5 -> " << solution.carPooling(trips, 5) << '\n';

    return 0;
}

File: 09-intervals/car-pooling/solution.go
package main

import (
	"fmt"
	"sort"
)

func carPooling(trips [][]int, capacity int) bool {
	events := make([][2]int, 0, len(trips)*2)

	for _, trip := range trips {
		passengers := trip[0]
		start := trip[1]
		end := trip[2]

		events = append(events, [2]int{start, passengers})
		events = append(events, [2]int{end, -passengers})
	}

	sort.Slice(events, func(i, j int) bool {
		if events[i][0] == events[j][0] {
			return events[i][1] < events[j][1]
		}
		return events[i][0] < events[j][0]
	})

	currentLoad := 0
	for _, event := range events {
		currentLoad += event[1]
		if currentLoad > capacity {
			return false
		}
	}

	return true
}

func main() {
	trips := [][]int{{2, 1, 5}, {3, 3, 7}}

	fmt.Printf("capacity = 4 -> %t\n", carPooling(trips, 4))
	fmt.Printf("capacity = 5 -> %t\n", carPooling(trips, 5))
}
